src/providers/aws.rs

```diff
@@ -0,0 +1,858 @@
+                Region::CaCentral1 => "ami-0bc1dd4eb012a451e",   // Canada
+                Region::EuCentral1 => "ami-0cdab515472ca0bac",   // Frankfurt
+                Region::EuNorth1 => "ami-c37bf0bd",              // Stockholm
+                Region::EuWest1 => "ami-01cca82393e531118",      // Ireland
+                Region::EuWest2 => "ami-0a7c91b6616d113b1",      // London
+                Region::EuWest3 => "ami-033e0056c336ecff0",      // Paris
+                Region::SaEast1 => "ami-094c359b4d8c6a8ca",      // Sao Paulo
+                Region::UsEast1 => "ami-064a0193585662d74",      // N Virginia
+                Region::UsEast2 => "ami-021b7b04f1ac696c2",      // Ohio
+                Region::UsWest1 => "ami-056d04da775d124d7",      // N California
+                Region::UsWest2 => "ami-09a3d8a7177216dcf",      // Oregon
+                x => panic!("Unsupported Region {:?}", x),
+            }
+            .into(),
+        )
+    }
+}
+
+impl Into<String> for UbuntuAmi {
+    fn into(self) -> String {
+        self.0
+    }
+}
+
+/// A descriptor for a particular machine setup in a tsunami.
+///
+/// The `setup` argument is called once for every spawned instances of this type with a handle
+/// to the target machine. Use [`Machine::ssh`](struct.Machine.html#structfield.ssh) to issue
+/// commands on the host in question.
+#[derive(Clone)]
+pub struct MachineSetup {
+    region: Region,
+    instance_type: String,
+    ami: String,
+    setup: Option<Arc<dyn Fn(&mut ssh::Session, &slog::Logger) -> Result<(), Error> + Send + Sync>>,
+}
+
+impl super::MachineSetup for MachineSetup {
+    type Region = String;
+
+    fn region(&self) -> Self::Region {
+        self.region.name().to_string()
+    }
+}
+
+impl PartialEq for MachineSetup {
+    fn eq(&self, other: &Self) -> bool {
+        self.ami == other.ami && self.instance_type == other.instance_type
+    }
+}
+
+impl Eq for MachineSetup {}
+
+impl std::hash::Hash for MachineSetup {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        self.instance_type.hash(state);
+        self.ami.hash(state);
+    }
+}
+
+impl Default for MachineSetup {
+    fn default() -> Self {
+        MachineSetup {
+            region: Region::UsEast1,
+            instance_type: "t3.small".into(),
+            ami: UbuntuAmi::from(Region::UsEast1).into(),
+            setup: None,
+        }
+    }
+}
+
+impl MachineSetup {
+    /// Set up the machine in a specific EC2
+    /// [`Region`](http://rusoto.github.io/rusoto/rusoto_core/region/enum.Region.html).
+    ///
+    /// The default region is us-east-1. [Available regions are listed
+    /// here](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/using-regions-availability-zones.html#concepts-available-regions)
+    /// AMIs are region-specific. This will overwrite the ami field to
+    /// the Ubuntu 18.04 LTS AMI in the selected region.
+    pub fn region(mut self, region: Region) -> Self {
```

Comment:
I think the general preference for builder-like types such as this one is to take and return `&mut`. See [`C-BUILDER`](https://rust-lang-nursery.github.io/api-guidelines/type-safety.html#c-builder).

Reply:
Related comment here: https://github.com/jonhoo/tsunami/pull/22/files#diff-60699702750aef5a8cdc050b4fce64eaR306

Reply:
ok, actually, using `&mut` makes this more annoying to use in general, as here: https://github.com/jonhoo/tsunami/pull/22/files#diff-46f9af62cc3650069927935c2a6dd4c8R63

a one-liner with a consuming builder has to get split into multiple lines with `&mut`. The current code is easier to write IMO.